### DISCUSSION TITLE
Version 11.6.0

### DISCUSSION THREAD

**Discussion by metastable-void** [menhera-org/TabArray#186] [Announcements]
## New features

- Drag and drop tabs inside a container to reorder them.
- Tab drag and drop between containers.
- New option to disable sorting of tabs.
- Group of containers (supergroups).
- Tags to organize tabs inside a single container.

## Improvements and fixes

- Adopted state of art tab sorting algorithm.
- New paths for add-on pages.
- Moved common codes into external libraries.
- Preparation work for a more responsible and performant non-persistent background page.
- Removed legacy codes.
- New language: Chinese (Simplified).
- Translation updates.
- Dependencies updates.
- Bug fixes.

Download and install from: https://addons.mozilla.org/firefox/addon/container-tab-groups/

<hr /><em>This discussion was created from the release <a href='https://github.com/menhera-org/TabArray/releases/tag/v11.6.0'>Version 11.6.0</a>.</em>

**Comment by JohnLGalt** [comment]
Hi there,

I'm not sure if I have an issue because of the addon, because of a recent Firefox update, or a combination of stuff, but up until now, I've been very successfully using your addon with Fx to get the best of all worlds that I wanted - tabs in the sidebar, separated by containers that are persistent across browser restarts, and a panorama view when I really want one.

However, either since a new Fx stable update (v112.0.2 64 bit from 25 April) or this addon's update (11.6.2, same 25 April date) I'm having containers randomly show up on start / while browsing with all tabs deleted. I've not added any new addons, and have been running the same profile for well over a year.

Any place that you'd like me to start? Would you rather I created an issue first and work from there, of is here OK?

**Reply by metastable-void** [reply]
Is is correct that you have browser sessions not correctly restored sometimes? That's strange, I think.

**Reply by metastable-void** [reply]
Firefox experts may advise cleaning up the profile when strange issues happen.

**Reply by metastable-void** [reply]
I have been using the new version of add-on, and yet to found anything strange related to session restoration.

**Reply by metastable-void** [reply]
How about console outputs? Open `about:debugging` > This Firefox and select Inspect for this extension.

**Reply by JohnLGalt** [reply]
I'll give the debugger a shot.

Interestingly, I use this addon in both Firefox stable with a dedicated profile as well as Firefox Nightly with a (different)  dedicated profile. Thus far, only the stable one has been acting up and not restoring / randomly deleting tabs from containers.

If I find something that breaks in a new profile with the addon only, then I will start a bug report. Until then, if I cannot narrow it down to the addon only, I won't.

**Reply by metastable-void** [reply]
What I thought of:

- Unintended interaction between addons: if the problem is this addon acting wrong, bug reports are welcome.
- Issues happening only with some configuration: do you have some settings changed in either Firefox or the addon?

**Comment by JohnLGalt** [comment]
It could be any of those, or more. As noted, both Firefox stable and this addon were updated on the same day.

My issues continue, but I haven't had a chance to look at the debugger yet, as I'm fighting to keep the tabs from disappearing (thank goodness for history). But as soon as I get a chance I'll post here.

**Reply by JohnLGalt** [reply]
One thing I just noticed, as it happened to me in Firefox nightly as well now, is that it occurs whenever I use the close button of a tab in a container - it's not that container that gets affected, though, it is a container that is not actually being used / is not focused.

Most of the containers I'm using are the default ones from the Firefox container sets, personal, shopping, etc.

Now that this is happening on both stable and nightly, I need to see if it might be an addon interaction thing (both browsers have relatively similar addon setups for the daily use profiles) or if it is something else.

**Reply by JohnLGalt** [reply]
Also, I checked my options in the app. The only thing I can see that is different from 'default' options is that I had Free memory of tabs after inactivity set to 1 hour, which I believe I initially set around 3-4 months ago, as I had Fx grabbing 7-8 GB of RAM (I have 128 GB RAM (yes, 128 GB RAM, not SSD space) installed, so not _too_ worried about it, but tried that to see if it would help.

I'm making a new profile now, but is there any way to make a backup of the tabs in the containers that can be imported into a new profile?

**Reply by metastable-void** [reply]
I do not remember any methods besides using an extension, to backup/import tabs.

**Reply by JohnLGalt** [reply]
It's definitely linked to deleting a tab from a focused container. In Firefox stable right now, typing posts in a forum in my Forums container. I've opened, then closed (removed) 3 different tabs to news articles, and every time I've done so another unfocused container has lost its contents completely.

I've got the debugger open, but not really sure what I need to look at. Nor how to export it so you can see what is going on.

However, there are a few of these errors that seem they might be related:

```
Uncaught SyntaxError: JSON.parse: unexpected end of data at line 1 column 1 of the JSON data
    renderContainerWithTabs PopupRenderer.ts:200
[PopupRenderer.ts:200:26](webpack://CTG/src/pages/popup-v2/legacy/PopupRenderer.ts)
    renderContainerWithTabs PopupRenderer.ts:200
```

```
Uncaught SyntaxError: JSON.parse: unexpected end of data at line 1 column 1 of the JSON data
    renderContainerWithTabs PopupRenderer.ts:226
[PopupRenderer.ts:226:24](webpack://CTG/src/pages/popup-v2/legacy/PopupRenderer.ts)
    renderContainerWithTabs PopupRenderer.ts:226
```

It's a set of 4, first an error with ts:200, then followed by an error with ts:226

ts:200 error leads to this line in the code:

`const data = JSON.parse(json);`

ts:226 leads to this line in the code:

` const data = JSON.parse(json);`

Same basic line, but it seems that it is not finding something it is looking for?

**Comment by JohnLGalt** [comment]
I just saw that you've released a new version. I'll give that a shot and see what happens.

**Reply by JohnLGalt** [reply]
New version hasn't exhibited the issue thus far. However, when focusing a different container, any open container seems to create a temporary tab with some sort of UUID before that container is unfocused. Opening the container, though, shows no temporary tab with a UUID name, so it is instantaneously created and showed right before the container loses focus.

I'll continue to monitor the current 11.7.0 build in both stable and nightly, and if I have any issues, I'll report them in a proper bug report, including info from the debugger.

**Reply by metastable-void** [reply]
Do you have any extension that changes the new tab page?

**Reply by JohnLGalt** [reply]
No, but I do have Tab Mix Plus, which changes things like link clicking behavior. To note, though, I've been using these profiles, both stable profile in stable Fx and nightly profile in Nightly Fx for many months now, alongside CTG addon, with no issue.

**Reply by JohnLGalt** [reply]
Also, Nightly profile just failed catastrophically.

I went to close a tab, and **every** container group was killed.

Here are the debugging CTG console logs, I've done one with only errors and debug, and one with every option enabled:
[Firefox.nightly.errorsanddebugonly.console-export-2023-5-3_11-3-37.txt](https://github.com/menhera-org/TabArray/files/11384049/Firefox.nightly.errorsanddebugonly.console-export-2023-5-3_11-3-37.txt)
[firefox.nightly.all.console-export-2023-5-3_11-4-15.txt](https://github.com/menhera-org/TabArray/files/11384050/firefox.nightly.all.console-export-2023-5-3_11-4-15.txt)

**Reply by JohnLGalt** [reply]
The main reason I opted to start using CTG, as noted above, is that it combined all the best features that I wanted:

Sidebar tabs
Container Groups
Persistent containers

I'm more than willing to do whatever I need to, based upon your suggestions, to test this out. I suspect this is a niche issue, either from other addon interaction, and / or a fundamental change in Firefox itself (though I think the reality of it being this latter case is extremely low, as with Nightly, it has the same feature set months in advance of the stable Firefox, being 2 major versions ahead, but both stable and nightly profiles started exhibiting the issue at exactly the same time, so if it was a Firefox issue, manifesting in Fx stable, it should have manifested in Fx Nightly months before). I'm good with making new profiles, as many as needed, in both stable and Nightly, to test, as well as any other thing you want me to do, just let me know. I would much rather continue using CTG, even if it means dropping an addon, and thus, here is my addon list for Nightly:

[Firefox.Nightly.Addons.txt](https://github.com/menhera-org/TabArray/files/11384101/Firefox.Nightly.Addons.txt)
If you'd like, I can paste a file with the entire text of about:support as well.

**Reply by metastable-void** [reply]
Thanks very much for using CTG, and I am very sorry to hear that sometimes tabs are lost. I am suspecting "Index Tab" feature is killing tabs, from console outputs.

**Reply by metastable-void** [reply]
What is your current setting for Index Tabs? Index Tabs feature is experimental and has been causing many bugs that lead to loss of tabs.

**Comment by JohnLGalt** [comment]
Let's call this done. I'm on 11.7.0, in a brand new profile, and everything not only works a lot faster than before, but with much less hassle as well.

I will create a real issue now if I continue to have issues in the brand new profile.

Happy Birthday!

**Reply by metastable-void** [reply]
I do not remember major changes that could solve the problem, so I am a bit afraid that it could happen again.

Here is the fulll change set from 11.6.2 to 11.7.0:

https://github.com/menhera-org/TabArray/compare/v11.6.2...v11.7.0

**Reply by JohnLGalt** [reply]
With new profiles in both Nightly and Stable, with 11.7.0, and with a much parsed down userChrome.css (hiding tabs across the top only), it seems to be working just fine again. I've had no instances of containers losing their tabs anymore.

I had a lot of customizations in userChrome.css, not just hiding the tab strip, but also several that had to do with sidebar. This is the file that I used in the previous Stable profile:

```
// Hide the horizontal tab strip
#TabsToolbar:not([customizing="true"]) {
  visibility: collapse !important;
}
/* if the titlebar is also hidden and your browser is maximized,
 * clicking on the top pixel of the screen should trigger menu items (Fitt’s law)
 */
#toolbar-menubar {
  padding-top: 0px !important;
}

/* taken from https://medium.com/@Aenon/firefox-hide-native-tabs-and-titlebar-f0b00bdbb88b */
/* hides the native tabs */
#TabsToolbar {
	visibility: collapse;
	}

// leaves space for the window buttons 
#nav-bar {
    margin-top: -8px;
    margin-right: 74px;
    margin-bottom: -4px;
	}


/*/ Sidebar border improvements for dark theme
#sidebar-header {
  background: #0C0C0D;
  border-bottom: none !important;
}
#sidebar-splitter {
  border-right-color: #0C0C0D !important;
  border-left-color: #0C0C0D !important;
}
#sidebar-switcher-target,
#sidebar-close {
  filter: invert(100%);
}*/

/*/ Floating auto-shrinking sidebar
#sidebar-box:not([hidden]) {
  display: block;
  position: fixed;
  min-width: 4vw;
  max-width: 4vw;
  overflow: hidden;
  transition: min-width 0.2s ease;
  border-right: 1px solid #cdc7c2;
  z-index:1;
}*/

/*#sidebar,
#sidebar-box:hover {
  min-width: 20vw !important;
  max-width: 20vw !important;
}

#sidebar-splitter {
  display: none;
}*/

/*
 * You need to subtract the height of the panels above sidebar: toolbar (with the URL),
 * bookmarks bar, etc. -- whichever is enabled/displayed.
 * You can see which mode (normal, compact or touch) you’re in by going to:
 * Firefox Menu → Customize… (at the bottom of the screen) → Density
 *
 *                | normal | compact | touch
 * Menu bar       |  27px  |   27px  |  27px
 * Tab bar        |  33px  |   29px  |  41px
 * Toolbar        |  40px  |   32px  |  40px
 * Bookmarks bar  |  23px  |   21px  |  27px
 * Sidebar header |  47px  |   47px  |  47px
 *
 * Ex: in compact mode:
 * - menu bar, tab bar and sidebar header hidden (0px)
 * - nav bar (32px) + bookmarks bar (21px) = 53px
 */
 
/*#sidebar {
  height: calc(100vh - 120px);
}*/

/*#sidebar-box:not([hidden]) ~ #appcontent {
  margin-left: 4vw;
}

#main-window[inFullscreen][inDOMFullscreen] #appcontent {
  margin-left: 0;
}

#main-window[inFullscreen] #sidebar {
  height: 100vh;
}*/

```

It's entirely possible that one or more of those settings, which I had incorporated for previous sidebar / tabs in side bar addons, may have been the root cause, as well.

At any rate, I do have one issue to report with 11.7.0 but I will make an issue out of it, as it is not a showstopper at all.

Hope you had a great birthday, and May the 4th be with you!

**Comment by metastable-void** [comment]
I found a problem with Firefox consuming a lot of memory and going unresponsive.

This is the bug report: https://bugzilla.mozilla.org/show_bug.cgi?id=1831463

**Comment by JohnLGalt** [comment]
I normally close out my browsers even if I leave my computer on, so this evening I will make another new profile, add only CTG, open several tabs in multiple containers and leave them open overnight to see if the memory leak is present here as well.

Additionally, I added myself to the bug CC and voted on it as well.

**Reply by metastable-void** [reply]
If the browser is still responsive, please open "Debugging Information" from three-dot button at the top of sidebar. You can see a performance report from there.

**Comment by JohnLGalt** [comment]
I was not home yesterday, but when I came back the browser was indeed responsive. In fact, it was using nominal memory. Unfortunately, I did not see your last post until just now, after I had closed it down and started all over.

I'll run another test with the current release version in one profile, and the new test version in another profile, and leave them both running overnight again. In the morning, I"ll grab the performance reports for both.

**Reply by metastable-void** [reply]
For the new version, "Save Debugging Information" button is available, so I would also like to see the downloaded JSON file.
